Chapter10/pretty_print.erl

```diff
@@ -0,0 +1,194 @@
+			end
+	end.
+
+
+
+
+
+
+
+
+
+% Exercise 10-1: Pretty-Printing
+prettyEntry({Word, Lines}) when length(Lines) == 1 ->
+	Paded = pad(20, Word),
+	[Entry] = Lines,
+
+	Output = Paded ++ char_to_integer(Entry),
+
+	io:format("~p~n",[Output]);
+
+prettyEntry({Word, Lines}) ->
+	Acc = accumulate(Lines),
+	PList = prettyList(Acc),
+
+	Paded = pad(20, Word),
+
+	Output = Paded ++ PList,
+
+	io:format("~p~n",[Output]).
+
+
+
+
+
+% This function should take a list of line numbers,
+% in descending order, and produce a list containing
+% ranges as well as removing duplicates.
+accumulate(List) ->
+	Sorted = lists:sort(List),
+	% Compressed = compress(Sorted).
+	compress_link(Sorted).
+
+
+% This function will print the output of accumulate so
+% that on the list [{1},{3}, {5,7}] the output is 1,3,5-7
+prettyList(List) ->
+	compress_pretty(List).
+
+
+% This function, called with number N and string Word,
+% will return the string padded to the right with spaces
+% to give it length N (assuming Word is not longer than N).
+pad(N, Word) when length(Word) < N -> pad(N, Word ++ [$\s]);
+pad(_, Word) -> Word.
+
+
+
+
+%%Compresser (Removes duplicates) [1,1,1,2,2,5,7,9,9] -> [1,2,5,7,9]
+compress([]) -> [];
+compress(List) -> compress(List, []).
+
+compress([], Output) -> Output;
+compress(List, Output) when length(List) == 1 -> Output ++ List;	
+compress([Head|Tail], _) when (Tail - Head) =< 1 -> [];
+compress([Head|Tail], Output) when (Tail - Head) >= 1 -> Output ++ [Head];
+compress([Head,Next|Tail], Output) ->
+	if
+		(Next - Head) == 0 -> compress([Head|Tail], Output);
+		% (Next - Head) == 1 -> compress([Next|Tail], Output ++ [Head] ++ ["-"] );
+		(Next - Head) >= 1 -> compress([Next|Tail], Output ++ [Head])
+	end.
+
+
+%%Linker (Links adjacent numbers) [1,2,5,8,9,10] -> [{1,2},{5},{8,10}]
+compress_link([]) -> [];
+compress_link(List) -> compress_link(List, {0, 0}, []).
+
+compress_link([], _, Output) -> Output;
+compress_link([Head,Next], Range, Output) -> 
+	if
+		(Next - Head == 1) and (Range == {0, 0}) -> Output ++ [{Head, Next}];
+		(Next - Head == 1) -> Output ++ [{element(1, Range), Next}];
+		(Next - Head >= 1) and (Range /= {0, 0}) -> Output ++ [{element(1, Range),Head},{Next}];
+		(Next - Head >= 1) -> Output ++ [{Head},{Next}]
+	end;
+compress_link([Head,Next|Tail], Range, Output) ->
+	if
+		(Next - Head == 1) and (Range == {0, 0}) -> compress_link([Next|Tail], {Head, 0}, Output);
+		(Next - Head == 1) -> compress_link([Next|Tail], {element(1, Range), Next}, Output);
+		(Next - Head >= 1) and (Range == {0, 0}) -> compress_link([Next|Tail], {0, 0}, Output ++ [{Head}]);
+		(Next - Head >= 1) -> compress_link([Next|Tail], {0, 0}, Output ++ [{element(1, Range),Head}])
+	end.
+
+
+%Prettyer (List with tuples to string) [{1,2},{3},{5,9}] -> "1-2,3,5-9,"
+compress_pretty([]) -> [];
+compress_pretty([Head|Tail]) when tuple_size(Head) == 1 -> char_to_integer(element(1,Head)) ++ "," ++ compress_pretty(Tail);
```

Comment:
Try to match on tuple, rather than use `tuple_size/1` to find out type of `Head`

Reply:
All fixed